[Python] Extends unit testing infra to support several python versions

vguerra opened the pull request:
In order to run the tests using different Python versions, two new
lit-directives were added: target-run-python2-interop-simple-swift and
target-run-python3-interop-simple-swift. Each of them behave exactly
like target-run-simple-swift only that they set the env variable
PYTHON_VERSION with the desired major python version to be loaded.

As well, within `python_runtime.swift` there is logic to check
if a given PYTHON_VERSION is loadable. All tests in the file
are short-circuited if the desired python library wasn't. Downside
is that this not play nice with lit failing error's reporting.

In this solution there is no duplication of logic in `lit.cfg`
to search for paths to different versions of the Python library
files as it only provides directives that set `PYTHON_VERSION`
before running tests.

~~This moves all tests present in test/Python/python_runtime.swift
into a private module within the stdlib: PythonUnittest, so that
they can be reused to run them against different Python versions.~~

~~In order to run the tests using different Python versions, two new
lit-directives were added: target-run-python2-interop-simple-swift and
target-run-python3-interop-simple-swift. Each of them behave exactly
like target-run-simple-swift only that they set the env variable
PYTHON_LIBRARY with the desired python library file path before
executing the produced test's binary. Now, lit.cfg has functionality
to look for python2 and python3 library file paths.~~

~~As well, two new lit-requirements were added: python2 and python3.
This way, tests for either version will be run only when the feature
is available.~~

----

Notes:

~~* I left the "Errors" test duplicated in python2_runtime and python3_runtime b/c when
moving it into the private module in the stdlib the compiler was crashing. I would prefer
to first merge this and then investigate the reason of the crash.~~
* In my local machine, the tests for python3 fail due to the checks for memory usage.
I raised the issue here https://github.com/apple/swift/pull/23316#issuecomment-476834100.
I would like to see if this tests fail as well in the CI when using python3.

Resolves [TF-336](https://bugs.swift.org/browse/TF-336).

rxwei commented on stdlib/private/PythonUnittest/PythonUnittest.swift:
```diff
@@ -0,0 +1,358 @@
+    
+    let versionPrefix = "\(major).\(minor)."
+
+    self.testWithLeakChecking("\(versionPrefix)CheckVersion") {
```
Why is it necessary to move tests to stdlib? That kind of makes these tests not tests anymore.

I was expecting a lightweight test utility that takes a bunch of versions that individual test cases can specify, not moving all tests elsewhere. Something like:

```swift
PythonRuntimeTestSuite.testWithLeakChecking("Range", versions: (2, 7), (3, 6)) {
  ...
}
```

dan-zheng replied:
I believe specifying Python-versions-to-test as an argument (e.g. `versions: (2, 7), (3, 6)`) is difficult to support, since dynamically switching Python version is not supported.

Please check out https://github.com/apple/swift/pull/23745#issuecomment-479246345.

rxwei replied:
I see. Thanks for the context. Moving tests to stdlib/private/ is definitely non-standard, and I like the option of using multiple RUN-lines. Do we have a JIRA issue about dynamic version switching?

dan-zheng replied:
After [some discussion](https://github.com/apple/swift/pull/23055#issuecomment-469418796), dynamically switching Python versions has been deemed a non-goal.

Pasting relevant context:

> > @pvieito: could you please confirm the behavior of `PythonLibrary.useVersion`? What are your thoughts on dynamically switching Python versions?
> 
> Loading two versions of `Python` in the same process is not possible, as a lot of common symbols would clash.
> 
> Switching Python versions (in the sense of unloading a Python library and loading other version) could be _technically possible_ in some OS (`dlclose`?), but would make the implementation a lot more complex and unstable (what happens when unloading Python 3 while holding a Python 3 `str` reference and then passing it to a Python 2 function?).

---

Filed [TF-415](https://bugs.swift.org/browse/TF-415) and closed it as `Won't fix`.

rxwei replied:
Yeah, symbol collision is a hard problem. However, if we are sure dynamic switching is not supported, `PythonLibrary.useVersion` should print an error when a version is already selected.

dan-zheng replied:
> However, if we are sure dynamic switching is not supported, `PythonLibrary.useVersion` should result in a precondition failure when a version is already selected.

That's a great suggestion! Let me file a bug.

vguerra replied:
@dan-zheng thanks for the explanations you gave before on this.  I'll add my comments on the other conversation you started.

and indeed ... TF-416 would make `PythonLibrary.useVersion` more robust.

dan-zheng commented:
Hi @vguerra,

Thanks for tackling this!

Unfortunately, I think I misled you by failing to update the [TF-336](https://bugs.swift.org/browse/TF-336) bug description: moving tests to a `PythonUnittest` module is not an ideal solution. Sorry about that!

---

Could you please focus on a solely `lit` based solution using `RUN` lines to run tests for both Python 2 and 3? It seems you have quite some `lit` work done already.

I wonder if we can achieve something better than requiring both Python 2 and 3 to run tests:
```swift
// RUN: %target-run-python2-interop-simple-swift
// RUN: %target-run-python3-interop-simple-swift
// REQUIRES: executable_test
// REQUIRES: python2
// REQUIRES: python3
```

I think you can instead bake the `REQUIRES: python(2|3)` requirements cleverly into the `RUN` line logic itself, so that `%target-run-python-interop-simple-swift` runs with Python 2 only if Python 2 is available, and runs with Python 3 only if Python 3 is available. Some more design is necessary - does the direction make sense to you?

dan-zheng commented on test/lit.cfg:
```diff
@@ -346,6 +346,35 @@ else:
+
+    try:
+        library_file_path = subprocess.check_output(
+            [exec_name, '-c', 
```
I'm not sure that `subprocess.check_output([exec_name, ...])` is a robust way to check that the Python 2/3 libraries are installed. Ideally, the logic should be consistent with the library searching logic in [`PythonLibrary.loadPythonLibrary()`](https://github.com/apple/swift/blob/e7f674e7a664fa2d35c1f5323a304e57e8b8cfd1/stdlib/public/Python/PythonLibrary.swift#L183).

Other reviewers: any ideas?

pschuh replied:
It isn't quite as clean in the test results, but I think you can use: PythonLibrary() in stdlib/public/Python. Pass in an explicit version, and then short-circuit all the tests if that version of python is not loadable. You have to be very careful to do this as the first thing of the program. In this model, you would blindly run the tests for all versions, but only the versions that exist would run. Thus, lit would not need to be aware of this.

vguerra commented:
@dan-zheng no worries :). I should have posted on the ticket earlier to verify the approach I was taken was the way to go.

Regarding the approach of two `RUN` statements and two `REQUIRES`: It basically works, the downside is that the two `RUN` statements wont be executed if either of the two features is not available. This approach would only work if we require python2 and python3 to be available on the host running the tests. People trying to run this tests would have to have installed both python versions, which I think it not ideal.

Baking the requirements into the a `%target-run-python-interop-simple-swift` substitution is, i think, doable. I will give this a try :).

vguerra committed:
[Python][TF-336] Extends unit testing infra to support several python versions

In order to run the tests using different Python versions, two new
lit-directives were added: target-run-python2-interop-simple-swift and
target-run-python3-interop-simple-swift. Each of them behave exactly
like target-run-simple-swift only that they set the env variable
PYTHON_VERSION with the desired major python version to be loaded.

As well, within `python_runtime.swift` there is logic to check
if a given PYTHON_VERSION is loadable. All tests in the file
are short-circuited if the desired python library wasn't. Downside
is that this not play nice with lit failing error's reporting.

In this solution there is no duplication of logic in `lit.cfg`
to search for paths to different versions of the Python library
files as it only provides directives that set `PYTHON_VERSION`
before running tests.

vguerra commented:
hello all!

I just pushed a new version of this following advice from @pschuh on this [comment](https://github.com/apple/swift/pull/23745#discussion_r271535654) but this is still WIP ( changed this PR title to reflect so ).

The main change being that the new lit directives only set the `PYTHON_VERSION` env var before launching the tests and leverages the python library loading logic within `PythonLibrary`.

@pschuh some questions:

* I was wondering how I can check if a given version of python is *loadble* or not. I extended `PythonLibrary` with `isPythonLibraryLoadable` which simply calls the private method `loadPythonLibrary()` to see if it returns nil. Is this acceptable or is there a better way of performing this check?
* As well, `init()` method of `PythonLibrary` does a `fatalError` and this constructor is used to initialize the `shared` property. So even if I provide the method I mentioned before, at some point this `fatalError` is triggered which interrupts execution of tests :(. I understand that throwing the `fatalError` is the safest thing to do here but would there be a way around it?.

Last thing, sorry for the delay on following this up but my time was a bit limited lately :/
